src/algebra/jordan/triple.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2022 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import algebra.lie.of_associative
+
+/-!
+# Jordan triples
+
+Let `A` be a module over a ring `R`. A triple product on `A` is a trilinear map
+$\{.\,.\,\}:A\times A\times A\mapsto A$ which is symmetric in the first and third variables. The
+module `A` is said to be a Jordan triple if, for any `a`, `b`, `c`, `d` and `e` in `A` the following
+Leibniz rule is satisfied:
+$$
+\{a\,b\,\{c\,d\,e\}\} = \{\{a\,b\,c\}\,d\,e\} - \{c\,\{b\,a\,d\}\,e\} + \{c\,d\,\{a\,b\,e\}\}
+$$
+A module `A` over a *-ring `R` is said to be a *-triple if it has a triple product linear and
+symmetric in the first and thrid variable and conjugate linear in the second variable. A *-triple
+satisfying the Leibniz rule is said to be a Jordan *-triple.
+
+As well as being of algebraic interest, Jordan *-triples arise naturally in mathematical physics,
+functional analysis and differential geometry. For more information about these connections the
+interested reader is referred to [alfsenshultz2003], [chu2012], [friedmanscarr2005],
+[iordanescu2003] and [upmeier1987].
+
+## Main results
+
+(None yet, we're just getting started.)
+
+## References
+
+* [Chu, Jordan Structures in Geometry and Analysis][chu2012]
+-/
+
+lemma biadd_expand {A : Type*} [add_comm_monoid A] (B: A →+ A →+ A) (a₁ a₂ b₁ b₂ : A) :
+  B (a₁ + a₂) (b₁+b₂) = B a₁ b₁ + B a₂ b₁ + B a₁ b₂ + B a₂ b₂ :=
+  by rw [map_add, map_add, add_monoid_hom.add_apply, add_monoid_hom.add_apply, ← add_assoc]
+
+/-- An additive commutative monoid with a trilinear triple product -/
+class has_trilinear_tp (A : Type*) [add_comm_monoid A] := (tp : A →+ A →+ A →+ A )
+
+notation ⦃a, b, c⦄ := has_trilinear_tp.tp a b c
+
+lemma add_left (A : Type*) [add_comm_monoid A] [has_trilinear_tp A] (a₁ a₂ b c : A) :
+  ⦃a₁ + a₂, b, c⦄ = ⦃a₁, b, c⦄ + ⦃a₂, b, c⦄ :=
+by rw [map_add, add_monoid_hom.add_apply, add_monoid_hom.add_apply]
+
+lemma add_middle (A : Type*) [add_comm_monoid A] [has_trilinear_tp A] (a b₁ b₂ c : A) :
+  ⦃a, b₁ + b₂, c⦄ = ⦃a, b₁, c⦄ + ⦃a, b₂, c⦄ := by rw [map_add, add_monoid_hom.add_apply]
```

Comment:
This is (or should be) `add_monoid_hom.map_add₂`

Reply:
I'm not sure there is a `add_monoid_hom.map_add₂`? Perhaps there should be? Did you mean  `linear_map.map_add₂`? https://github.com/leanprover-community/mathlib/blob/e37daad5a168786e066050a8fe1d7f065f652c90/src/linear_algebra/bilinear_map.lean#L124

Reply:
Something along these lines, perhaps?
```lean
class is_jordan_triple (S : Type u) [comm_semiring S] [star_ring S]
(A : Type v) [add_comm_group A] [module S A]
  extends is_jordan_tp A :=
(triple_smul₁ : ∀ (s : S) (a b c : A), ⦃s•a, b, c⦄ = s • ⦃a, b, c⦄)
(triple_smul₂ : ∀ (s : S) (a b c : A), ⦃a, s•b, c⦄ = (star s) • ⦃a, b, c⦄)

variables (S : Type u) [comm_semiring S] [star_ring S]
variables {A : Type v} [add_comm_group A] [module S A]

def sesterlinear [is_jordan_triple S A] : A →ₗ[S] A →ₛₗ[star_ring_end S] A →ₗ[S] A := {
  to_fun := λ a, {
    to_fun := λ b, {
      to_fun := λ c, ⦃a, b, c⦄,
      map_add' := λ c₁ c₂, sorry,
      map_smul' := sorry,
    },
    map_add' := sorry,
    map_smul':= sorry,
  },
  map_add' := λ a₁ a₂, sorry,
  map_smul' := sorry,
}
```

Reply:
Trying something like this:
```lean
import algebra.lie.of_associative

universes u v

/-- An additive commutative monoid with a trilinear triple product -/
class has_triadditive_tp (A : Type*) [add_comm_group A] := (tp : A →+ A →+ A →+ A )

localized "notation ⦃a, b, c⦄ := has_triadditive_tp.tp a b c" in triple

class has_sesterlinear_tp (R : Type u) [comm_semiring R] [star_ring R] (A : Type*)
[add_comm_group A] [module R A] :=
(tp : A →ₗ[R] A →ₛₗ[star_ring_end R] A →ₗ[R] A )

instance : star_ring ℤ := star_ring_of_comm


lemma add_left {A : Type*} [add_comm_group A] [has_triadditive_tp A] (a₁ a₂ b c : A) :
  ⦃a₁ + a₂, b, c⦄ = ⦃a₁, b, c⦄ + ⦃a₂, b, c⦄ :=
by rw [map_add, add_monoid_hom.add_apply, add_monoid_hom.add_apply]

instance has_triadditive_tp_to_has_sesterlinear_tp {A : Type v} [add_comm_group A] 
[has_triadditive_tp A] :  has_sesterlinear_tp ℤ A := {
  tp := {
    to_fun := λ a, {
    to_fun := λ b, {
      to_fun := λ c, ⦃a, b, c⦄,
      map_add' := λ c₁ c₂, sorry,
      map_smul' := sorry,
    },
    map_add' := sorry,
    map_smul':= sorry,
  },
  map_add' := λ a₁ a₂, λ b, linear_map.ext (λ c, add_left a₁ a₂ b c)
  map_smul' := sorry,
  }
}
```
But it doesn't quite meet up in the middle
```lean
mvp.lean:33:28
type mismatch at application
  linear_map.ext (λ (c : A), add_left a₁ a₂ b c)
term
  λ (c : A), add_left a₁ a₂ b c
has type
  ∀ (c : A),
    ⇑(⇑(⇑has_triadditive_tp.tp (a₁ + a₂)) b) c =
      ⇑(⇑(⇑has_triadditive_tp.tp a₁) b) c + ⇑(⇑(⇑has_triadditive_tp.tp a₂) b) c
but is expected to have type
  ∀ (x : A), ⇑?m_10 x = ⇑?m_11 x
```

Reply:
There's a discrepancy here that my brain can't quite unpick at the moment. `→+` is a `add_monoid_hom`, whereas `→ₛₗ[star_ring_end R]` is an instance of `add_monoid_hom_class`. So it is not quite right in lean to say that  `→ₛₗ[star_ring_end R]` is an instance of `→+` - which seems a bit unsatisfactory.

Reply:
That's the point of `add_monoid_hom_class`: you don't have to use `add_monoid_hom` to talk about additive monoid homs.

Reply:
So why is the definition of `→+`
```
infixr ` →+ `:25 := add_monoid_hom
```
and not
```
infixr ` →+ `:25 := add_monoid_hom_class
```
?

Reply:
`add_monoid_hom A B` is **the** type of additive monoid homomorphisms from `A` to `B`. `add_monoid_hom_class F A B` states that an **already existing** type `F` is **a** type of additive monoid homomorphisms from `A` to `B`. So for example `add_monoid_hom` is an `add_monoid_hom_class`, but so is `add_equiv`, `ring_hom`, `continuous_add_monoid_hom`...

Put another way, `add_monoid_hom` is here to be studied as an object on its own, by giving it an algebraic structure, an order, a topology... while `add_monoid_hom_class` is a way to talk about additive monoid homs individually.